File: 0077. 组合/回溯.js
// 128ms
var combine = function(n, k) {
    function combination(start, k) {
        let result = []
        if (k == 1) {
            for (let i = start; i <= n; i += 1) {
                result.push([i])
            }
        } else {
            for (let i = start; i <= n; i += 1) {
                if (n + 1 - i >= k) {
                    let pres = combination(i + 1, k - 1)
                    for (let pre in pres) {
                        pres[pre].unshift(i)
                        result.push(pres[pre])
                    }
                }
            }
        }
        return result
    }
    return combination(1, k)
};

File: 0018. 四数之和/回溯.swift
// 3500ms
class Solution {
    func fourSum(_ nums: [Int], _ target: Int) -> [[Int]] {
        var nums = nums.sorted()
        func search(_ idx: Int, _ target: Int, _ num: Int) -> [[Int]] {
            if num == 1 {
                for i in idx..<nums.count {
                    if nums[i] == target {
                        return [[target]]
                    }
                }
                return [[Int]]()
            } else {
                var result = [[Int]](), pre = Int.max
                for i in idx..<nums.count {
                    var current = nums[i]
                    if current == pre {
                        continue
                    } else {
                        pre = current
                        var pres = search(i + 1, target - current, num - 1)
                        pres.forEach { pre in
                            result.append([current] + pre)
                        }
                    }
                }
                return result
            }
        }
        return search(0, target, 4)
    }
}